File: ros_ws/src/helpers/utcoupe_metrics/src_py/utcoupe_metrics/metric_types/chronometer.py
#!/usr/bin/env python

__author__ = "Gaëtan Blond"
__date__ = "20/03/20"

import rospy

from gauge import Gauge

class Chronometer(object):
    def __init__(self, topic_name):
        self.m_gauge = Gauge(topic_name)
        self.m_startTime = None
    
    def start(self):
        self.m_startTime = rospy.get_rostime()
    
    def stop(self):
        now = rospy.get_rostime()
        if self.m_startTime is None:
            raise Exception("must call start before stop")
        dur = now - self.m_startTime
        self.m_gauge.set(dur.to_nsec())
        self.reset()
    
    def reset(self):
        self.m_startTime = None

File: ros_ws/src/helpers/utcoupe_metrics/src/utcoupe_metrics/metric_types/chronometer.cpp
#include "utcoupe_metrics/metric_types/chronometer.h"

#include <exception>

void utcoupe_metrics::metric_types::Chronometer::publish()
{
    m_gauge.publish();
}


void utcoupe_metrics::metric_types::Chronometer::start() {
    m_started = true;
    m_startTime = ros::Time::now();
}

void utcoupe_metrics::metric_types::Chronometer::stop()
{
    if (!m_started) {
        throw std::runtime_error("chronometer hasn't been started");
    }
    auto deltaT = ros::Time::now() - m_startTime;
    m_gauge.set(static_cast<double>(deltaT.toNSec()));
}
